File: Lecture2/data_visualization/script/data_structures.R
# This script is for discovering, generating and editing data structures in R

# 1. Scalar 
sc <- 1 # Define a scalar (single value) and examine its length and type
length(sc) # Get length of scalar
class(sc) # Get class of scalar

# 2. Vector
## Numeric vectors
a <- c(1, 2, 3, 4, 5:10) # Create a numeric vector using the combine function c()
a # Display the vector
length(a) # Display the length of the vector
class(a) # Display the class (type) of the vector
str(a) # Use the str() function to compactly display the structure of the vector

## Character vectors
b <- c("A", "B", "C", "D") # Create character vector
b # Display the vector
length(b) # Display the length of the vector
class(b) # Display the class (type) of the vector
str(b) # Display the structure of the vector

## Combining vectors
e <- c(a, b) # Combine the numeric vector 'a' and character vector 'b' into a new vector 'e'
e # Display the combined vector

# 3. Matrix
my_mat <- matrix(1:16, nrow = 4, byrow = TRUE) # Create a 4x4 matrix with elements 1 to 16 filled by row
my_mat[1,] # Access and display the first row of the matrix

my_array <- array(1:16, dim = c(2, 4, 2)) # Create a 3-dimensional array with dimensions 2x4x2
my_array # Display the array

## Add colnames and rownames
rownames(my_mat) <- c("A", "B", "C", "D") # Set the row names 
colnames(my_mat) <- c("a", "b", "c", "d") # Set the column names 
my_mat # Display the matrix with row names and column names

# 4. List 
# Create a list with 3 elements: a character vector, a logical vector, and a matrix
list_1 <- list(
    c("black", "yellow", "orange"),
    c(TRUE, TRUE, FALSE, TRUE, FALSE, FALSE),
    matrix(1:6, nrow = 3)
)
# Display the list
list_1

# Create a named list
list_2 <- list(
    colours = c("black", "yellow", "orange"),
    evaluation = c(TRUE, TRUE, FALSE, TRUE, FALSE, FALSE),
    time = matrix(1:6, nrow = 3)
)
list_2 # Display the named list

# 5. Data frame
# Create vectors for height, weight, and names
p.height <- c(180, 155, 160, 167, 181)
p.weight <- c(65, 50, 52, 58, 70)
p.names <- c("Joanna", "Charlotte", "Helen", "Karen", "Amy")

# Create a data frame using these vectors
dataf <- data.frame(height = p.height, weight = p.weight, names = p.names)
dataf # Display the data frame

## Discover dataframe 
dim(dataf) # Show the dimensions of the data frame (number of rows and columns)
str(dataf) # Show the structure of the data frame
colnames(dataf) # Show the column names of the data frame
dataf[1,] # Show the first row of the data frame

# 6. Remove NA
# Create a dataframe with some NA (missing) values
Name <- c("Tuan", "Nhan", "Ngoc", "Tien", "Kien")
Height <- c(1.63, 1.78, 1.60, 1.56, NA)
Weight <- c(60, 72, 58, NA, NA)
df <- data.frame(Name, Height, Weight)

df # Display the dataframe
na.omit(df) # Remove rows with any NA values
complete.cases(df) # Return a logical vector indicating which cases are complete (no NA values)
df[complete.cases(df), ] # Use logical indexing to show only complete cases
table(is.na(df)) # Count the number of NA and non-NA values in the dataframe
df[rowSums(is.na(df)) == 0, ] # Use logical indexing to show rows with no NA values
rowSums(is.na(df)) == 0 # Return a logical vector where TRUE indicates rows with no NA values

File: Lecture2/data_visualization/script/data_structures.py
# This script is for discovering, generating, and editing data structures in Python

# Section 1: Scalar
sc = 1  # Define a scalar
print(sc)  # Print the scalar

# Section 2: List (Python's equivalent of R's vector)
a = list(range(1, 11))  # Create a numeric list
print(a)  # Display the list
print(len(a))  # Display the length of the list
print(type(a))  # Display the type of the list

b = ["A", "B", "C", "D"]  # Create a character list
print(b)  # Display the list
print(len(b))  # Display the length of the list
print(type(b))  # Display the type of the list

e = a + b  # Combine the numeric list 'a' and character list 'b' into a new list 'e'
print(e)  # Display the combined list

# Section 3: 2-Dimensional List (Python's equivalent of R's matrix) & 3-Dimensional List (Python's equivalent of R's array)
import numpy as np

my_mat = np.array(range(1, 17)).reshape(4, 4)  # Create a 4x4 matrix
print(my_mat[0,:])  # Access and display the first row of the matrix

my_array = np.array(range(1, 17)).reshape(2, 4, 2)  # Create a 3-dimensional array
print(my_array)  # Display the array

# Section 4: Dictionary (Python's equivalent of R's list)
list_1 = {
    "colours": ["black", "yellow", "orange"],
    "evaluation": [True, True, False, True, False, False],
    "time": np.array(range(1, 7)).reshape(3, 2)
}  # Create a dictionary
print(list_1)  # Display the dictionary

# Section 5: DataFrame (Python's equivalent of R's dataframe)
import pandas as pd

p_height = [180, 155, 160, 167, 181]  # Create a list for height
p_weight = [65, 50, 52, 58, 70]  # Create a list for weight
p_names = ["Joanna", "Charlotte", "Helen", "Karen", "Amy"]  # Create a list for names

dataf = pd.DataFrame({"height": p_height, "weight": p_weight, "names": p_names})  # Create a dataframe using the lists
print(dataf)  # Display the dataframe

print(dataf.shape)  # Show the dimensions of the dataframe
print(dataf.info())  # Show the structure of the dataframe
print(dataf.columns)  # Show the column names of the dataframe
print(dataf.iloc[0,:])  # Show the first row of the dataframe

# Section 6: Handle Missing Values
Name = ["Tuan", "Nhan", "Ngoc", "Tien", "Kien"]  # Create a list for names
Height = [1.63, 1.78, 1.60, 1.56, None]  # Create a list for height with a missing value
Weight = [60, 72, 58, None, None]  # Create a list for weight with missing values

df = pd.DataFrame({"Name": Name, "Height": Height, "Weight": Weight})  # Create a dataframe with missing values
print(df)  # Display the dataframe

print(df.dropna())  # Remove rows with any missing values
print(df.dropna().shape[0] == df.shape[0])  # Check if all cases are complete
print(df[~df.isna().any(axis=1)])  # Display only complete cases
print(df.isna().sum())  # Count the number of missing values
print(df[~df.isna().any(axis=1)])  # Display rows with no missing values
print(df.isna().sum(axis=1) == 0)  # Get a boolean series indicating rows with no missing values
